File: leetcode/239.sliding-window-maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#

from typing import List, Tuple
from common import BaseTestCase, unittest
# @lc code=start
import heapq


class PriorityQueue:
    def __init__(self):
        self.nums: List[int] = []
        
    def push(self, data, priority=0):
        heapq.heappush(self.nums, (priority, data)) 

    def pop(self) -> Tuple[int, any]:
        return heapq.heappop(self.nums)

    def top(self) -> Tuple[int, any]:
        if self.nums:
            return self.nums[0]

    def empty(self) -> bool:
        return len(self.nums) == 0


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        pq = PriorityQueue()
        for r in range(k-1):
            pq.push(r, priority=(-1)*nums[r])

        ans = [0]*(n-k+1)
        l = 0
        for r in range(k-1, n):
            pq.push(r, priority=(-1)*nums[r])
            while not pq.empty() and pq.top()[1] < l:
                pq.pop()

            _, idx = pq.top()
            ans[l] = nums[idx]
            l += 1
        
        return ans            
        
# @lc code=end
class P239TestCase(BaseTestCase):
    def test_1(self):
        nums = [1,3,-1,-3,5,3,6,7]
        k = 3
        expected = [3,3,5,5,6,7]
        actual = Solution().maxSlidingWindow(nums, k)
        self.assertEqual(expected, actual)

    def test_2(self):
        nums, k = [1], 1
        actual = Solution().maxSlidingWindow(nums, k)
        self.assertEqual([1], actual)

    def test_3(self):
        nums, k = [1,3,1,2,0,5], 3
        actual = Solution().maxSlidingWindow(nums, k)
        expected = [3,3,2,5]
        self.assertEqual(expected, actual)
    
    
if __name__ == "__main__":
    unittest.main()

File: leetcode/239.sliding-window-maximum.cpp
/*
 * @lc app=leetcode id=239 lang=cpp
 *
 * [239] Sliding Window Maximum
 */
#include <iostream>
#include <queue>
#include <tuple>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
   public:
    // https://leetcode.com/problems/sliding-window-maximum/solutions/65898/clean-c-o-n-solution-using-a-deque/
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> ans;
        for (int i = 0; i < nums.size(); i++) {
            if (!dq.empty() && dq.front() == i - k) 
                dq.pop_front();

            while (!dq.empty() && nums[dq.back()] < nums[i])
                dq.pop_back();

            dq.push_back(i);
            if (i >= k - 1) 
                ans.push_back(nums[dq.front()]);
        }
        return ans;
    }

    // O(nlogn) solution using max heap
    vector<int> maxSlidingWindow_maxHeap(vector<int>& nums, int k) {
        priority_queue<tuple<int, int>> pq;
        for (int i = 0; i < k - 1; ++i) {
            pq.push({nums[i], i});
        }

        vector<int> ans(nums.size() - k + 1, 0);
        for (int r = k - 1; r < nums.size(); ++r) {
            pq.push({nums[r], r});
            int l = r - k + 1;

            while (pq.empty() == false && get<1>(pq.top()) < l) {
                pq.pop();
            }

            ans[l] = get<0>(pq.top());
        }

        return ans;
    }
};
// @lc code=end
int main() {
    // vector<int> nums{1, 3, -1, -3, 5, 3, 6, 7};
    vector<int> nums{1, 3, 1, 2, 0, 5};

    Solution sl;
    // vector<int> ans = sl.maxSlidingWindow(nums, 3);
    vector<int> ans = sl.maxSlidingWindow_dequeu(nums, 3);
    // vector<int> ans = sl.maxSlidingWindow_maxHeap(nums, 3);

    cout << "len: " << ans.size() << endl;
    for (int i = 0; i < ans.size(); i++) {
        cout << ans[i] << " ";
    }
    cout << endl;
}
